Sources/NIO/Codec.swift

```diff
@@ -101,115 +129,266 @@ private extension ChannelHandlerContext {
+        /// Yes please, here we go.
+        case available(ByteBuffer)
+    }
 
-            precondition(slice.writerIndex == sliceWriterIndex, "Writing to the buffer is not allowed")
+    /// Result of a try to process a buffer.
+    enum BufferProcessingResult {
+        /// Could not process a buffer because we are already processing one on the same call stack.
+        case cannotProcessReentrantly
+        /// Yes, we did process some.
+        case didProcess(DecodingState)
+    }
 
-            self.cumulationBuffer!.moveReaderIndex(forwardBy: sliceReadable - slice.readableBytes)
+    private var state: State = .ready
+    private var buffers: CircularBuffer<ByteBuffer> = CircularBuffer(initialRingCapacity: 4)
+}
 
-            // If the user told us more data is needed we also need to ensure the writerIndex did not change in between.
-            // If the writerIndex changed we need to retry as there is more data delivered via re-entrance maybe.
-            if result == .needMoreData && self.cumulationBuffer!.writerIndex == writerIndex {
-                break
+// MARK: B2MDBuffer Main API
+extension B2MDBuffer {
+    /// Start processing some bytes if possible, if we receive a returned buffer (through `.available(ByteBuffer)`)
+    /// we _must_ indicate the processing has finished by calling `finishProcessing`.
+    mutating func startProcessing() -> BufferAvailability {
+        switch self.state {
+        case .processingInProgress:
+            return .bufferAlreadyBeingProcessed
+        case .ready:
+            if self.buffers.count > 0 {
+                var buffer = self.buffers.removeFirst()
+                mergeBuffers(self.buffers, into: &buffer)
+                self.buffers.removeAll(keepingCapacity: true)
+                if buffer.readableBytes > 0 {
+                    self.state = .processingInProgress
+                    return .available(buffer)
+                } else {
+                    return .nothingAvailable
+                }
+            } else {
+                return .nothingAvailable
             }
         }
     }
 
-    /// Calls `decode` until there is nothing left to decode.
-    public func channelRead(ctx: ChannelHandlerContext, data: NIOAny) {
-        // Either merge the received data into the existing cumulationBuffer or use it as the cumulationBuffer if none exists yet.
-        if self.cumulationBuffer != nil {
-            var buffer = self.unwrapInboundIn(data)
-            self.cumulationBuffer!.write(buffer: &buffer)
+
+    mutating func finishProcessing(remainder buffer: ByteBuffer) -> Void {
+        assert(self.state == .processingInProgress)
+        self.state = .ready
+        if buffer.readableBytes > 0 {
+            self.buffers.prepend(buffer)
         } else {
-            self.cumulationBuffer = self.unwrapInboundIn(data)
+            var buffer = buffer
+            buffer.clear()
+            mergeBuffers(self.buffers, into: &buffer)
```

Comment:
Should there ever be any buffers in `self.buffers` at the time of this call?

Reply:
@Lukasa yes, if we re-entrantly got some more. So for example:

    startProcessing
        decode
            write
                pipeline.fireChannelRead (because draining the socket buffer on error
    finishProcessing